### ISSUE TITLE
support locating multiple configuration files

### ISSUE THREAD

**Issue by tkellen** [js-cli/js-liftoff#8]
configuration like this:
```js
new Liftoff({
  configFiles: {
    '.app': {
      extensions: ['rc','.yml'],
      search: {
        cwd: {
          path: './',
          findUp: true
        },
        myHomeKey: '~',
        etc: '/etc'
      }
    },
    'appfile': {
      extensions: function () {
        return Object.keys(require.extensions)
      },
      searchPaths: {
        baseDir: {
          path: './',
          findUp: true
        },
        home: '~',
        alternate: '/path/to/alternate',
        multiple: ['/path/to/whatever', '/path/to/something', '/path/to/something/else']
      }
    }
  }
});
```

yields environment like this:
```js
env.configFiles = {
  '.app': {
    cwd: '/var/www/project/code/.apprc',
    myHomeKey: '/Users/user/.apprc',
    etc: null
  },
  'appfile': {
    baseDir: '/var/www/project/code/appfile.js',
    home: null,
    alternate: null,
    multiple: ['/path/to/whatever/appfile.js', '/path/to/something/appfile.js']
  }
};
```

`~` expands to `process.env.HOME || process.env.HOMEPATH || process.env.USERPROFILE`
`searchPath` entries which specify an array of paths should NOT perform findUp on each entry.

**Comment by sttk** [comment]
@tkellen I tried this issue and finished implementing it on my repository.

[sttk/js-liftoff#support_multiple_config_files](https://github.com/js-cli/js-liftoff/compare/master...sttk:support_multiple_config_files)

But I changed a bit of your specification:

* The properties `searchPaths` or `searchPath` are not necessary.
* If the `searchPath` property is specified, the result is a string or null. Otherwise the result is always an array.
* If the `extensions` property is an object, this module gets an sorted array of its property keys and uses it.

**Comment by sttk** [comment]
For the example in the top comment, the program outputs as follows:

```js
(/Users/sttk/foo/bar/liftoff/tester.js)
var Liftoff = require('.');

var _require = {
  extensions: {
    '.js': null,
    '.json': null,
  },
};

var Tester = new Liftoff({
  name: 'app',
  configFiles: {
    '.app': {
      extensions: ['rc', '.yml'],
      searchPaths: {
        cwd: {
          path: './',
          findUp: true,
        },
        myHomeKey: '~',
        etc: '/etc',
      },
    },
    'appfile': {
      extensions: function() {
        return Object.keys(_require.extensions);
      },
      searchPaths: {
        baseDir: {
          path: './',
          findUp: true,
        },
        home: '~',
        alternate: '/tmp/path/to/alternate',
        multiple: ['/tmp/path/to/whatever', '/tmp/path/to/something', '/tmp/path/to/something/else'],
      },
    }
  },
});

Tester.launch({}, function(env) {
  var util = require('util');
  console.log(util.inspect(env, { depth: null }));
});
```
```
(output)
{ cwd: '/Users/sttk/foo/bar/liftoff',
  require: [],
  configNameSearch: [ 'appfile.js', 'appfile.json' ],
  configPath: null,
  configBase: undefined,
  modulePath: undefined,
  modulePackage: {},
  configFiles: 
   { '.app': 
      { cwd: [ '/Users/sttk/foo/bar/liftoff/.apprc' ],
        myHomeKey: [ '/Users/sttk/.apprc' ],
        etc: [] },
     appfile: 
      { baseDir: [],
        home: [],
        alternate: [],
        multiple: 
         [ '/tmp/path/to/whatever/appfile.js',
           '/tmp/path/to/something/appfile.js' ] } } }
```

**Comment by sttk** [comment]
@tkellen Thanks for giving me push access, but I want you to check whether my modification meets your specification before commit this to the repository.

And now, I'm considering that searched paths should be returned not like `'searched/file/path.json'` but like `{ extension: '.json', path:'searched/file/path.json' }` for applying require functions corresponding to their extensions after.
How do you think about it?

**Comment by tusbar** [comment]
@sttk could you open a PR here so we can review the changes properly? :)

**Comment by sttk** [comment]
@tusbar I've open this PR. Please review.

**Comment by phated** [comment]
Fully fleshed out idea from https://github.com/js-cli/js-liftoff/pull/67#discussion_r59125527

Possible new spec:
```js
new Liftoff({
  // extensions are already supported by liftoff
  extensions: {
    '.js': null,
    '.yml': 'yml-loader-module'
  },
  configFiles: {
    '.app': {
      // value can be an object with `path`, `findUp` and `extensions` properties
      cwd: {
        path: './',
        findUp: true,
        // explicit extensions that override top-level extensions
        // extension overrides would be handled exactly the same as top-level extensions
        // everything else defaults to top-level extensions
        extensions: {
          'rc': 'some-special-rc-loader'
        }
      },
      // value can be a string that expands to `{ path: value }`
      // if extensions aren't passed using an object, top-level extensions are used
      myHomeKey: '~',
      etc: '/etc'
    },
    'appfile': {
      baseDir: {
        path: './',
        findUp: true
      },
      home: '~',
      alternate: '/path/to/alternate',
      // value can also be an array of strings or objects (strings get expanded)
      multiple: ['/path/to/whatever', '/path/to/something', '/path/to/something/else']
    }
  }
});
```
yields
```js
env.configFiles = {
  '.app': {
    cwd: '/var/www/project/code/.apprc',
    myHomeKey: '/Users/user/.app.yml',
    etc: null
  },
  'appfile': {
    baseDir: '/var/www/project/code/appfile.js',
    home: null,
    alternate: null,
    multiple: ['/path/to/whatever/appfile.js', '/path/to/something/appfile.js']
  }
};
```

/cc @tkellen @sttk

**Comment by sttk** [comment]
I understand new spec as follows (and have some questions) :

1. `searchPaths/searchPath` properties are abolished.
   * **Question**: If multiple paths are hit with extensions, is the result an array of paths, or only a first path?
   * **Question**: If the result path can be a string or an array, do users have to check it before using it?
2. Each `configFiles` properties (e.g. `.app`) can be an object.
   * **Question**: Does it support a string and an array? _(it is supported now.)_
3. Each properties of `configFiles` properties (e.g. `cwd` and `myHomeKey`) can be an object, a string, a string array, or an object array. If an object (including an element of an array), `path` property is mandatory in it.
4. `configFiles` property (e.g. `.app`) is used as a part of path.
   * **Question**: Is it allowed to include `'.'` and `'~'` characters? _(it is allowed now)_

**Comment by phated** [comment]
Thinking deeper about your questions, I think every result should be a singular string.  That would mean dropping the array syntax (and relying on the keyed object for fallbacks) or stopping traversal after the first match; which do you think is better?  I'm leaning towards dropping the array syntax.

Extensions: match the behavior that liftoff currently uses.  If there is a gulpfile.js and a gulpfile.babel.js in a project, which is used?

I don't think `configFiles` should accept an array or string and should only accept an object.

The properties of `configFiles` are supposed to be the file's basename, as it is currently written.  If we wanted to do away with that and use generic keys (like `rc`, etc), we would have to add a `name` property to the objects and I don't really like that.  I'm on the fence about using the key as the file's basename but it is probably for the best.  If it stays that way, all characters are part of the filename (and are appended with the extensions), nothing in the key is expanded (e.g. `~` is not expanded to home) and only valid path characters should be allowed as keys.

Hope that helps, but it might open up other questions.

**Comment by sttk** [comment]
@phated Thank you for answering my questions.

> That would mean dropping the array syntax ... or stopping traversal after the first match; which do you think is better? 

If every result is a singular string, I also think that dropping the array syntax is better because it makes less confusing for users and I can't think the cases in which the syntax is used.

> ... If there is a gulpfile.js and a gulpfile.babel.js in a project, which is used?

Liftoff preferentially uses an earlier extension in specified order. (Though it isn't guaranteed.)

> I don't think configFiles should accept an array or string and should only accept an object.

I got it. I'll do so.

> ... nothing in the key is expanded (e.g. ~ is not expanded to home) and only valid path characters should be allowed as keys.

I got it. I'll implement not to expanded '~' and '.' but to use the key just as user specify.

**Comment by phated** [comment]
Support for this was just published as 2.3.0
